cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/type/DatasetTypeManager.java

```diff
@@ -393,6 +419,24 @@ private void deployDefaultModules() {
     }
   }
 
+  private void deployExtensionModules() {
+    // adding any defined extension modules to be available in dataset manager service
+    for (Map.Entry<String, DatasetModule> module : extensionModules.entrySet()) {
+      try {
+        // NOTE: we assume extension modules are always in classpath, hence passing null for jar location
+        // NOTE: we add extension modules in the system namespace
+        Id.DatasetModule theModule = Id.DatasetModule.from(Id.Namespace.SYSTEM, module.getKey());
+        addModule(theModule, module.getValue().getClass().getName(), null);
+      } catch (DatasetModuleConflictException e) {
+        // perfectly fine: we need to add the modules only the very first time service is started
+        LOG.info("Not adding " + module.getKey() + " extension module: it already exists");
+      } catch (Throwable th) {
+        LOG.error("Failed to add {} extension module. Aborting.", module.getKey(), th);
```

Comment:
Sorry for the confusion. For the log and rethrow case, it's the inconsistency we have in older code base. Recently we established a convention on not logging and rethrow. See http://wiki.cask.co/display/CE/Infrastructure+Improvements+-+3.2

For the case that addition exception message need to be added to the exception going to be thrown, it's ok to create a new exception with a modified message, but the original `Throwable` cause must be preserved. E.g.

```java
try {
...
} catch (SomeException e) {
  throw new RuntimeException("Some message", e);
}
```